README.md
# Hate Group DNS Blacklist

- [Overview](#overview)
- [Files & Notes](#files--notes)
	- [Pi-hole](#pi-hole)
	- [ISC BIND](#isc-bind)
	- [BlueCat Address Manager/Response Policies](#bluecat-address-managerresponse-policies)
- [To do](#to-do)
- [Sources](#sources)
- [License](#license)

## Overview
This is a list of designated hate groups' domains as designated by the [Southern Poverty Law Center](http://splcenter.org). Data is pulled from their [Hate Map](https://splcenter.org/hate-map)

Please report errors [here](https://github.com/chigh/hategroup-dnsbl/issues).

**Note:** This project and I am are in *no* way associated with SPLC. I created this with data from their public website plus a handful of other domains I have come across. (see the Custom sections in *import.csv*)

- Not all groups have domains; Some have facebook or pages on other shared resources. This will be noted in the *import.csv* if further information is available. This information will be periodically back-filled as it becomes available.
- Some ideologies have overlap. Pi-hole will account for that. 

Included in this bundle are files and configuration code to be used with [Pi-hole](https://pi-hole.net), [ISC BIND](https://isc.org) Response Policies, or [BlueCat](https://bluecatnetworks.com) Response Policies.

## Files & Notes 
- *block_hate.sh*
	- Script to generate the list objects
- *import.csv*
	- A comma separated variable file of domains and the names of the organizations. This is for reference and to generate all of the lists.
    - Lines beginning with 
        - `?` are groups/domains that are not in the list, but are on the SPLC site.
        - `??` domain needs to be confirmed, needs more detail, etc.
        - `+` are newly added groups in the list.
        - `-` are active groups which have been in previous year's lists, but are not in the current year's list (remain in configs)
        - `--` scheduled to be deleted with the next year's data
        - `nx` are domains which no longer exist (nxdomain) (removed from configs)
        - `d` are domains which still exist, but are non-functional (remain in configs)
        - `rc` are reclaimed domains for other purposes (removed from configs)

### Pi-hole
With Pi-hole, use either the wildcard list or the blacklist, but not both. 

- *blocklist.txt*
	- A text file list of only the domains.
	- This can be added to Pi-hole configuration on a local web server or add the following URL to Pi-hole's Block Lists ([instructions](https://discourse.pi-hole.net/t/how-do-i-add-additional-block-lists-to-pi-hole/259)):
		- https://raw.githubusercontent.com/chigh/hategroup-dnsbl/master/blocklist.txt
- *04-pihole-wildcard.conf*
	- Wildcard domain list for Pi-hole. When using the GUI, *03-pihole-wildcard.conf* is created and used. Used as a separate file, *dnsmasq* will read in the contents for both without having to worry about overwriting anything added within the GUI. Less work.
    - Modify this file to contain the IP address of your Pi-Hole appliance.
	- Copy this file into */etc/dnsmasq.d* onto your Pi-hole appliance and restart Pi-hole: 
		- `pihole restartdns`

### ISC BIND
- *hategrp.rpz.db*
	- BIND zone database file to be used with the contents of named.conf for BIND's Response Policies
- *named.conf*
	- A section of named.conf to be included in your BIND implementation to use the included database. This goes in the *view* section.

### BlueCat Address Manager/Response Policies
Importing the generated file will over-write all entries in the Response Policy in which this list is imported.

- *hategrp.rpz_BAM-import.txt*
	- A file to be imported into BlueCat Address Manager to create a Response Policy. ***Note*** Importing these objects into a Response Policy will completely overwrite any existing records.

## Sources 

| [Ideology] & [Sources]            | Date     |
| :--                               | :--:     |
| [Alt-Right]                       |[N/A]     |
| [Anti-Immigrant]                  |2020-06-17|
| [Anti-LGBTQ]                      |2020-06-17|
| [Anti-Muslim]                     |2020-06-17|
| [Antigovernment Movement]         |[N/A]     |
| [Black Separatist]  		        |2020-06-17|
| [Christian Identity] 		        |2020-06-17|
| [General Hate]                    |2020-06-17|
| [Hate Music]                      |2020-06-17|
| [Holocaust Denial]                |2020-06-17|
| [KKK]                             |2020-06-27|
| [Male Supremacy]                  |2020-06-17|
| [Neo-Confederate]                 |2019-08-10|
| [Neo-Nazi]                        |2021-01-11|
| [Neo-V&ouml;lkisch][neo-volkisch] |2021-01-11|
| [Phineas Priesthood]              |[N/A]     |
| [Racist Skinheads]                |2021-01-19|
| [Radical Traditional Catholicism] |2021-01-19|
| [Sovereign Citizens Movement]     |[N/A]     |
| [White Nationalist]               |2021-01-19|

- There is no SPLC for Canada, so based upon a groups related to the ones already in the list.
	- [A guide to Canada's anti-Islamic groups -- NowToronto](https://nowtoronto.com/news/canada-islamophobic-groups/)
- Also using [Center for Countering Digital Hate](https://counterhate.co.uk)


## To do

- Complete the Antigovernment Movement.
	- As of 2018, there were 475 groups listed. Some may be grouped under a single domain, but until I can research, they are all considered to be individual groups with potentially individual domains.
    - These are not included in the hatemap and have to be researched via the SPLC website

## License
Licensed under [MIT](https://raw.githubusercontent.com/chigh/hategroup-dnsbl/master/LICENSE.md).
[the nuts and bolts of putting this together]

[N/A]:  # "No specific groups are listed or found for this ideology."
[next]: # "This is next on the list to be done."
[sources]: https://www.splcenter.org/fighting-hate
[ideology]: https://www.splcenter.org/fighting-hate/extremist-files/ideology
[alt-right]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/alt-right
[anti-immigrant]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/anti-immigrant
[anti-lgbtq]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/anti-lgbt
[anti-muslim]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/anti-muslim
[antigovernment movement]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/antigovernment
[black separatist]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/black-separatist
[christian identity]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/christian-identity
[general hate]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/general-hate
[hate music]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/hate-music
[holocaust denial]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/holocaust-denial
[kkk]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/ku-klux-klan
[male supremacy]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/male-supremacy
[neo-confederate]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/neo-confederate
[neo-nazi]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/neo-nazi
[neo-volkisch]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/neo-volkisch
[phineas priesthood]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/phineas-priesthood
[racist skinheads]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/racist-skinhead
[radical traditional catholicism]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/radical-traditional-catholicism
[sovereign citizens movement]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/sovereign-citizens-movement
[white nationalist]: https://www.splcenter.org/fighting-hate/extremist-files/ideology/white-nationalist



CHANGELOG.md
## 2021-11-24
- ADDED: Gender Critical Coming Out Day -- Anti-LGBT+

## 2021-03-21
- ADDED: Potential domain for Mike Lindell's new free speech social network

## 2021-02-05

- ADDED: Hate comics sites

## 2021-01-19

- UPDATE: Racist Skinheads, Rad Trad Catholicism, White Nationalist w/2019 data
- FIXED: made sure all lines fit a "standard" with proper fields
- FIXED: typographical errors in non-domain fields
- ADDED: a couple of new and/or updated statuses. 
- CHANGED: re-sorted entries. Groups found on the SPLC website, but not on the hate-map, are now listed at the bottom of each category.
- PROPOSED: add 3rd date for when removed or leave last modified field for that
- PROPOSED: add additional fields to existing SPLC csv; would cause more work than to keep it separate

## 2021-01-11

- UPDATE: Neo-Nazi, Neo-Voelkisch

## 2020-11-14

- ADDED: two custom domains for rw militias and ultra right-wing media & conspiracy theories

## 2020-06-27

- FIXED: commented out some rows which were meant to be exactly that.
- ADDED: Graham Linehan's, known transphobe, blog domain
- UPDATE: KKK

## 2020-06-25

- ADDED: Alt-right social media websites

## 2020-06-17

- UPDATE: Started update with 2019 data. Anti-immigrant, Male Supremacist, and Wikipedia related
- UPDATE: Anti-LGBTQ, Anti-Muslim, Black Separatist, Christian Identity, General Hate, Hate Music, Holocaust Denial

## 2020-06-16

- ADDED: Center for Countering Digital Hate

## 2019-11-05

- ADDED: 8chan rebrand 8kun. Probably already offline, but blocking nonetheless.



block_hate.sh
#!/usr/bin/env bash
###########################################################################
## ID: block_hate.sh
## DESC: creates DNSBL for use with Pi-hole, BIND, and BlueCat
## AUTHOR: Clair High <github.com@fwd.tch3.com>
## DATE: 2017-10-13, 2019-08-10
###########################################################################
scratch=$(mktemp -d -t tmp.XXXXXXXXXX) # $scratch will be auto-deleted on
finish () { rm -rf $scratch ; }        # exit no matter the exit status
trap finish EXIT INT
###########################################################################
# SHELL SETTINGS
set -o errexit
set -o errtrace
set -o nounset
set -o pipefail
set -o posix
#set -o verbose
#set -o xtrace
###########################################################################
# USER CONFIGURABLE VARIABLES
prefix="hategrp"
input="import.csv" 
address="127.0.0.1"
###########################################################################
# PLEASE DO NOT CHANGE THE FOLLOWING
rpz_zone="${prefix}.rpz"
import_file="${rpz_zone}_BAM-import.txt" 
bind_db="${rpz_zone}.db"
pihole_txt="blocklist.txt"
input_tmp="${scratch}/${input}.$$" 
pihole_wildcard="04-pihole-wildcard.conf"
########################################################################### 

if [[ ! -f ${input} ]]
then
    printf "error: missing input file\n"
    exit 1
fi

domains=$(grep -v ^# ${input} | awk -F'|' '{print $2}' | sort)

function help() {
printf "Usage: $(basename $0) -abchin\n"
printf "Options:
    -a, --all     create RPZ import file for Address Manager and BIND database
    -b, --bind    create BIND database
    -c, --clean   clean all files
    -h, --help    prints this help message
    -i, --import  creates RPZ import file for Address Manager
    -n, --named   prints requisite named.conf info
    -p, --pihole  creates file that can be imported into Pi-hole\n"
}

function create_import_file() {
    if [[ -f $import_file ]]
        then rm ${import_file}
    fi
    for domain in ${domains[@]}
    do
        printf "**.${domain}\n" >> ${import_file}
    done
}

function create_pihole_list() {
    for file in ${pihole_txt} ${pihole_wildcard}
    do
        if [[ -e ${file} ]]
            then rm ${file}
        fi
    done
    
cat <<EOF >> ${pihole_txt}
# Hate group DNS Blocklist
# Based off Southern Poverty Law Center's list of hate groups. 
#   Source: https://www.splcenter.org/fighting-hate
#           https://www.splcenter.org/hate-map
#   URL: https://raw.githubusercontent.com/chigh/hategroup-dnsbl/master/blocklist.txt
# Date: $(date +%Y-%m-%d)
#
EOF

cat <<EOF >> ${pihole_wildcard}
# Hate group DNS Blocklist
# Based off Southern Poverty Law Center's list of hate groups. 
#   Source: https://www.splcenter.org/fighting-hate
#           https://www.splcenter.org/hate-map
#   URL: https://tch3.net/hg-dnsbl/blocklist.txt
#   URL: https://raw.githubusercontent.com/chigh/hategroup-dnsbl/master/blocklist.txt
#
# Copy this file to /etc/dnsmasq.d on your Pi-hole appliance and restart Pi-hole:
#
#   pihole restartdns
#
# Date: $(date +%Y-%m-%d)
#
EOF

    for domain in ${domains[@]}
      do 
        printf "${domain}\n" >> ${pihole_txt}
        printf "address=/${domain}/${address}\n" >> ${pihole_wildcard}
    done
}

function create_named_conf()  { 
    printf "Include the following in your named.conf in the DNS view\n\n"
    #printf  "   response-policy { zone \"${rpz_zone}\" policy CNAME localhost; };
    printf  "   response-policy { zone \"${rpz_zone}\" policy NODATA; };

    // zone: ${rpz_zone}
    zone \"${rpz_zone}\"
    {
        type master;
        file \"${bind_db}\";
        allow-transfer { 127.0.0.1; };
    };\n"
}

function create_bind_db() {
#sort -k2 -n ${input} > ${input_tmp}
cat > ${bind_db} <<EOL
\$TTL 5
@ IN SOA LOCALHOST. postmaster.no.email.please. ( 1 3600 600 2592000 3600 )
@ IN NS LOCALHOST.
LOCALHOST. IN A 127.0.0.1
EOL

    for domain in ${domains[@]}
    do
        printf "*.${domain} CNAME .\n" >> ${bind_db}
        printf "${domain} CNAME .\n" >> ${bind_db}
    done
}

function clean_up() {
    if [[ -f ${input_tmp} ]]
      then rm -f ${input_tmp}
    fi
}

function clean_all() {
    for file in ${input_tmp} ${bind_db} ${import_file}
    do
        if [[ -f ${file} ]]
          then rm -f ${file}
        fi
    done
}

case "${1-}" in
    -a|--all) create_named_conf && create_import_file && create_bind_db && create_pihole_list;;
    -b|--bind) create_bind_db      ;;
    -c|--clean) clean_all && exit 0 ;;
    -h|--help) help && exit 0      ;;
    -i|--import) create_import_file  ;;
    -n|--named) create_named_conf   ;;
    -p|--pihole) create_pihole_list ;;
     *) clean_up; printf "error: invalid option\n\n" && help ; exit 1 ;;
esac

clean_up
